beacon-chain/forkchoice/doubly-linked-tree/store.go

```diff
@@ -84,21 +88,50 @@ func (s *Store) head(ctx context.Context) ([32]byte, error) {
 	if bestDescendant == nil {
 		bestDescendant = justifiedNode
 	}
-	currentEpoch := slots.EpochsSinceGenesis(time.Unix(int64(s.genesisTime), 0))
+	currentSlot := slots.SinceGenesis(time.Unix(int64(s.genesisTime), 0))
+	currentEpoch := slots.ToEpoch(currentSlot)
 	if !bestDescendant.viableForHead(s.justifiedCheckpoint.Epoch, s.finalizedCheckpoint.Epoch, currentEpoch) {
 		s.allTipsAreInvalid = true
 		return [32]byte{}, fmt.Errorf("head at slot %d with weight %d is not eligible, finalizedEpoch, justified Epoch %d, %d != %d, %d",
 			bestDescendant.slot, bestDescendant.weight/10e9, bestDescendant.finalizedEpoch, bestDescendant.justifiedEpoch, s.finalizedCheckpoint.Epoch, s.justifiedCheckpoint.Epoch)
 	}
 	s.allTipsAreInvalid = false
 
+	secondsIntoSlot := (uint64(time.Now().Unix()) - s.genesisTime) % params.BeaconConfig().SecondsPerSlot
+	earlyCall := secondsIntoSlot >= params.BeaconConfig().ProcessAttestationsThreshold
+	currentlyOrphaning := s.proposerHeadNode != s.headNode
+	noBlock := s.highestReceivedNode.slot+1 < currentSlot || (earlyCall && s.highestReceivedNode.slot < currentSlot)
+	earlyBlock := s.highestReceivedNode.secondsSinceSlotStart(s.genesisTime) <= params.BeaconConfig().OrphanLateBlockFirstThreshold && s.highestReceivedNode.slot == currentSlot
+	if noBlock || earlyBlock {
+		if currentlyOrphaning {
+			currentlyOrphaning = false
+			orphanBetMisses.Inc()
+		}
+		s.proposerHeadNode = bestDescendant
+	} else if s.committeeBalance > 0 {
+		liveChain := s.proposerHeadNode.balance*100/s.committeeBalance > liveFraction
+		votedFraction := s.highestReceivedNode.balance * 100 / s.committeeBalance
```

Comment:
Also ideal to compute the votedFraction as an independent, pure function to reduce complexity here